### ISSUE TITLE
[java] False positives for "JUnit tests should include assert() or fail()"

### ISSUE THREAD

**Issue by benmccann** [pmd/pmd#1455]
**Affects PMD Version:** 6.4.0

**Rule:** JUnit tests should include assert() or fail()
[JUnitTestsShouldIncludeAssert](https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#junittestsshouldincludeassert)

**Description:** A large number of issues for our codebase are false positives generated by this rule. See https://app.codacy.com/app/jtablesaw/tablesaw/issues and select the rule under "Pattern"

This is because it PMD not detect asserts that are called indirectly by calling another method first

**Code Sample demonstrating the issue:**

https://github.com/jtablesaw/tablesaw/blob/ed5cedfa9884133160d43638c56683ca4964e7c1/core/src/test/java/tech/tablesaw/columns/times/PackedLocalTimeTest.java#L170

**Running PMD through:** Codacy

**Comment by matthieu-vergne** [comment]
Since the usual vocabulary is about "asserting" stuff, I would suggest to rename your `check` method to align with this convention. In particular, naming it `assertEquivalentTimes` would not only solve the warning, it would also be more explicit.

**Comment by rdohms** [comment]
What about for cases like this:
```
        mvc.perform(
                post("/endpoint/".concat("not-an-uuid"))
                .contentType(MediaType.APPLICATION_JSON))
                .andExpect(MockMvcResultMatchers.status().isBadRequest());
```

Is there a plugin i could be using? I'm sorry, i'm new to this and looking for resources.

**Comment by matthieu-vergne** [comment]
The rule assumes a somehow standard assertion. If you use a specific testing framework, which uses its own terminology, you just can't rely on generic rules like this one. So either you tweak your framework to follow the rule, either you discard the rule. A generic rule cannot afford to deal with any possible specific variants.

Nothing stops you from making your own specific rule, though.

**Comment by Fernal73** [comment]
A similar flagging of this rule happens when you use the EqualsVerifier.verify method which throws an AssertionError when it fails.

**Comment by firegloves** [comment]
I'm experiencing the same problem but I can't solve it by renaming the assertions method to something containing the word assert. Sometimes it works, for some others, it won't work. For example with this code, I get the issue:

```
@Test
public void columnSmallint() {
   this.assertOnMempoiColumn("column_SMALLINT", Types.SMALLINT, EExportDataType.INT);
}
```

I don't want to disable the rule, because it is useful in various cases.

**Comment by adangel** [comment]
Maybe the rule is confused by using "this" - have you tried without? e.g.:

```java
@Test
public void columnSmallint() {
   assertOnMempoiColumn("column_SMALLINT", Types.SMALLINT, EExportDataType.INT);
}
```

**Comment by adangel** [comment]
`AssertionHelper.assertOnSomething` is not supported, because the classname is not "Assert". It should work with statically importing it, since the method name starts with "assert", e.g.

```java
import static AssertionHelper.assertOnSomething;
public class Foo {
    @Test
    public void testSomething() {
        assertOnSomething(expected, actual);
    }
}
```

The corresponding rule code is here:
https://github.com/pmd/pmd/blob/81739da5caff948dbcd2136c17532b65c726c781/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitTestsShouldIncludeAssertRule.java#L162-L166

**Comment by ajeans** [comment]
@adangel I am getting bitten by this one too.

Both JUnit5 and AssertJ now use `Assertions.assert*` and `Assertions.fail`.

Would you consider a MR that simply adds those cases to the assertion detection?

 https://github.com/pmd/pmd/blob/81739da5caff948dbcd2136c17532b65c726c781/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitTestsShouldIncludeAssertRule.java#L162-L166 ?

**Comment by oowekyala** [comment]
Note that in pmd 7 the rule recognizes more cases as "asserts", see 

https://github.com/pmd/pmd/blob/423949cc276df6c6eaf312ef4ffbcb977fe258b1/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/TestFrameworksUtil.java#L130-L138

It would probably make sense to backport this to master.

**Comment by jsotuyod** [comment]
This is fixed in PMD 7.0.0-rc1. It won't be backported to PMD 6.
